tests/api/printf.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright 2023 The clvk authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef CLVK_UNIT_TESTING_ENABLED
+
+#include "testcl.hpp"
+#include "unit.hpp"
+#include "utils.hpp"
+
+#include <filesystem>
+
+#ifdef __APPLE__
+#include <unistd.h>
+#endif
+
+#ifdef WIN32
+#include <Windows.h>
+#include <io.h>
+#endif
+
+static std::string stdoutFileName;
+
+#define BUFFER_SIZE 1024
+static char stdoutBuffer[BUFFER_SIZE];
+
+static void releaseStdout(int fd) {
+    fflush(stdout);
+    dup2(fd, fileno(stdout));
+    close(fd);
+}
+
+static bool getStdout(int& fd) {
+    fd = dup(fileno(stdout));
+    if (!freopen(stdoutFileName.c_str(), "w", stdout)) {
+        fprintf(stderr, "ERROR!\n");
+        releaseStdout(fd);
+        return false;
+    }
+    return true;
+}
+
+static char* getStdoutContent() {
+    FILE* f;
+    memset(stdoutBuffer, 0, BUFFER_SIZE);
+    fflush(stdout);
+    f = fopen(stdoutFileName.c_str(), "r");
+    if (f == nullptr)
+        return nullptr;
+
+    char* ptr = stdoutBuffer;
+    do {
+        ptr += strlen(ptr);
+        ptr = fgets(ptr, BUFFER_SIZE, f);
+    } while (ptr != nullptr);
+    fclose(f);
+
+    return stdoutBuffer;
+}
```

Comment:
We should introduce a callback to get access to the output of printf from unit tests. As discussed, I want to make this standard in a future version of OpenCL, similarly to how we've done it in the cl_arm_printf extension.